nucypher/characters/lawful.py

```diff
@@ -548,43 +549,56 @@ def join_policy(self, label, alice_pubkey_sig, node_list=None, block=False):
         treasure_map = self.get_treasure_map(alice_pubkey_sig, label)
         self.follow_treasure_map(treasure_map=treasure_map, block=block)
 
-    def retrieve(self, message_kit, data_source, alice_verifying_key, label):
+    def retrieve(self, message_kit, data_source, alice_verifying_key, label, cache=False):
 
         capsule = message_kit.capsule  # TODO: generalize for WorkOrders with more than one capsule
-        capsule.set_correctness_keys(
-            delegating=data_source.policy_pubkey,
-            receiving=self.public_keys(DecryptingPower),
-            verifying=alice_verifying_key)
 
         hrac, map_id = self.construct_hrac_and_map_id(alice_verifying_key, label)
         _unknown_ursulas, _known_ursulas, m = self.follow_treasure_map(map_id=map_id, block=True)
 
-        # TODO: Consider blocking until map is done being followed.
+        already_retrieved = len(message_kit.capsule._attached_cfrags) >= m
```

Comment:
Why follow the TMap here - Is it possible to push this check earlier in this method?

Reply:
That's interesting.  It of course implies that we do have (or at least have the possibility of having) cached cfrags.  Wanna make an Issue?

Reply:
Following up : <https://github.com/nucypher/nucypher/issues/934>